### PR TITLE
F5 blueprint

### PR DESCRIPTION
**Pull request by LucaPrete**
The blueprint allows users to deploy active/active F5 Big-IP VE based LBs, both for internal (L4 ILBs) and external (L4 NLBs) use.

---
**Checklist**

I applicable, I acknowledge that I have:
- [x] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [x] Ran `terraform fmt` on all modified files
- [x] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [x] Made sure all relevant tests pass

### CODE REVIEW

File: blueprints/networking/f5-lb/load_balancers.tf [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  ilb_forwarding_rules_config = {
+    for k, v in var.forwarding_rules_config
+    : k => {
+      address       = try(google_compute_address.ilb_ip_addresses[k].ip_address, null)
+      global_access = v.global_access
+      ip_version    = v.ip_version
+      protocol      = v.protocol
+    } if v.external == false
+  }
+
```
Comment (ludoo) [inline blueprints/networking/f5-lb/load_balancers.tf:27]: nit: no blank lines between locals

File: blueprints/networking/f5-lb/load_balancers.tf [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    } if v.external == true
+  }
+
+  nlb_forwarding_rules_ip_addresses = {
+    for k, v in var.forwarding_rules_config
+    : k => v if v.address != null
+  }
+}
+
+resource "google_compute_address" "ilb_ip_addresses" {
```
Comment (ludoo) [inline blueprints/networking/f5-lb/load_balancers.tf:48]: can you use the `net-address` module instead of resources?

File: blueprints/networking/f5-lb/main.tf [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  # F5 configs
+  as3_url  = "https://github.com/F5Networks/f5-appsvcs-extension/releases/download/v3.46.0/f5-appsvcs-3.46.0-5.noarch.rpm"
```
Comment (ludoo) [inline blueprints/networking/f5-lb/main.tf:19]: this is super ugly, can we find a way to make it not hurt readers' eyes? I can give it a shot if you want; probably moving these in the actual startup script, then tokenizing what you pass in and doing the final manipulations in the template vars?

File: blueprints/networking/f5-lb/main.tf [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  do_url   = "https://github.com/F5Networks/f5-declarative-onboarding/releases/download/v1.39.0/f5-declarative-onboarding-1.39.0-4.noarch.rpm"
+  do_ver   = format("v%s", split("-", split("/", local.do_url)[length(split("/", local.do_url)) - 1])[3])
+  fast_url = "https://github.com/F5Networks/f5-appsvcs-templates/releases/download/v1.25.0/f5-appsvcs-templates-1.25.0-1.noarch.rpm"
+  fast_ver = format("v%s", split("-", split("/", local.fast_url)[length(split("/", local.fast_url)) - 1])[3])
+  init_url = "https://cdn.f5.com/product/cloudsolutions/f5-bigip-runtime-init/v1.6.2/dist/f5-bigip-runtime-init-1.6.2-1.gz.run"
+  ts_ver   = format("v%s", split("-", split("/", local.ts_url)[length(split("/", local.ts_url)) - 1])[2])
+  ts_url   = "https://github.com/F5Networks/f5-telemetry-streaming/releases/download/v1.33.0/f5-telemetry-1.33.0-1.noarch.rpm"
+}
+
+resource "google_compute_address" "dataplane_ip_addresses" {
```
Comment (ludoo) [inline blueprints/networking/f5-lb/main.tf:32]: here too I would use `net-address`

File: blueprints/networking/f5-lb/main.tf [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  subnetwork   = var.network_self_links["mgmt"]["subnetwork"]
+
+  address = (
+    var.f5_ip_addresses_config[each.key]["mgmt"] == null
+    ? null
+    : var.f5_ip_addresses_config[each.key]["mgmt"]
+  )
+}
+
+module "bigip-vms" {
```
Comment (ludoo) [inline blueprints/networking/f5-lb/main.tf:61]: you can have this module create instance groups so you avoid the two resources for groups

File: blueprints/networking/f5-lb/variables.tf [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    ip_version    = optional(string, "IPV4")
+    protocol      = optional(string, "L3_DEFAULT")
+  }))
+  description = "The optional configurations of the GCP load balancers forwarding rules."
+  default = {
+    "" = {}
+  }
+}
+
+variable "health_check_config" {
```
Comment (ludoo) [inline blueprints/networking/f5-lb/variables.tf:37]: if you pass this directly to the module, I would use `any` in the type

File: blueprints/networking/f5-lb/variables.tf [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+ * limitations under the License.
+ */
+
+variable "disk_size" {
+  type        = number
+  description = "The default disk size in GB for F5 VMs."
+  default     = 100
+}
+
+variable "forwarding_rules_config" {
```
Comment (ludoo) [inline blueprints/networking/f5-lb/variables.tf:23]: here too I would not expose the module-level variable, let's just do what makes sense here

File: blueprints/networking/f5-lb/variables.tf [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  description = "The machine type to deploy (i.e. n2-standard-8)."
+  default     = "n2-standard-4"
+}
+
+variable "name" {
+  type        = string
+  description = "The unique name to assign to the resources."
+}
+
+variable "network_self_links" {
```
Comment (ludoo) [inline blueprints/networking/f5-lb/variables.tf:134]: `vpc_config` maybe?

File: blueprints/networking/f5-lb/variables.tf [hunk 9]
```diff
@@ -0,0 +1,208 @@
+/**
```
Comment (ludoo) [inline blueprints/networking/f5-lb/variables.tf:1]: there are too many variables here, can we

- remove what is not essential and easy to change in the code (e.g. the load balancer configurations)
- group logically in `object`-typed variables (e.g. `instance_configs`, `f5_config`, etc)

**Review by ludoo — CHANGES_REQUESTED** [review]
very nice, dropped a few comments but happy to discuss f2f or do the changes together of course

File: blueprints/networking/f5-lb/main.tf (hunk 3 above)
Comment (LucaPrete) [reply blueprints/networking/f5-lb/main.tf:19]: I know... this is the only way I have to cleanly use the original script from F5 and possibly update it over time.
If I start hardcoding these in the script I'll have to redo the whole work every time I update it. Can we keep things here as they are?

Comment (LucaPrete) [reply blueprints/networking/f5-lb/main.tf:19]: think that in the original F5 scripts these values were directly injected in the VMs metadata :(

Comment (ludoo) [reply blueprints/networking/f5-lb/main.tf:19]: I see your point. I still think this might be beautified a bit, if you don't mind I can give it a try once everything else is ready.

Comment (LucaPrete) [reply blueprints/networking/f5-lb/main.tf:19]: Sure thing!

**Comment by LucaPrete** [comment]
Adding a message to bump the visibility of this :)

File: blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+      alias_ip_range_address = string
+      alias_ip_range_name    = string
+      dataplane_address      = optional(string)
+      management_address     = optional(string)
+    })
+    license_key = optional(string, "AAAAA-BBBBB-CCCCC-DDDDD-EEEEEEE")
+  }))
+}
+
+variable "f5_vms_shared_config" {
```
Comment (ludoo) [inline blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf:30]: this is not used in examples AFAIK, can we add one to show how this work and test it?

File: blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "f5_vms_dedicated_config" {
+  description = "The F5 VMs configuration. The map keys are the zones where the VMs are deployed."
+  type = map(object({
```
Comment (ludoo) [inline blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf:19]: nit: I understand the naming you used but don't love it :)

This variable holds the network configuration for VMs, can we rename as `instance_vpc_configs` or somesuch? and the one below as `instance_vm_config`?

TL;DR I would like variable names to make it immediately obvious that one configures network-related attributes for *each* VM, the other the underlying instance configuration for all.

If you really like this approach though, we could slightly change `f5_vms_shared_config` to `instance_shared_config`, and use `instance_vpc_configs` for the other.

WDYT?

File: blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    service_account = optional(string)
+    ssh_public_key  = optional(string, "my_key.pub")
+    tags            = optional(list(string), [])
+    use_gcp_secret  = optional(bool, false)
+    username        = optional(string, "admin")
+  })
+  default = {}
+}
+
+variable "forwarding_rules_config" {
```
Comment (ludoo) [inline blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf:47]: `_configs`

**Review by ludoo — CHANGES_REQUESTED** [review]
Very nice, dropped a few comments mainly on naming.

File: blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf (hunk 11 above)
Comment (LucaPrete) [reply blueprints/third-party-solutions/f5-bigip/f5-bigip-ha-active/variables.tf:19]: resolved in 1:1

**Review by ludoo — APPROVED** [review]
Luca, we can tweak it after merge. When you're fine with it go ahead and merge.
